src/App.tsx

```diff
@@ -1,223 +1,237 @@
+import React, {
+  useContext,
+  useEffect,
+  useRef,
+  useState,
+  FormEvent,
+  useMemo,
+  useCallback,
+} from 'react';
+import classnames from 'classnames';
+import {
+  createTodo,
+  getTodos,
+  removeTodo,
+  updateTodo,
+} from './api/todos';
 import { AuthContext } from './components/Auth/AuthContext';
+import { ErrorNotification } from './components/ErrorNotification';
+import { Footer } from './components/Footer';
+import { TodosList } from './components/TodosList';
+import { FilterType } from './types/FilterStatus';
+import { Todo } from './types/Todo';
+import { Error } from './types/Error';
 
 export const App: React.FC = () => {
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
   const user = useContext(AuthContext);
-  const newTodoField = useRef<HTMLInputElement>(null);
+  const inputField = useRef<HTMLInputElement>(null);
+  const [error, setError] = useState<Error | null>(null);
+  const [filterType, setFilterType] = useState<FilterType>(FilterType.All);
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [title, setTitle] = useState('');
+  const [isAdding, setIsAdding] = useState(false);
+  const [selectedTodos, setSelectedTodos] = useState<number[]>([]);
+  const [toggle, setToggle] = useState(true);
+  const [selectedTodo, setSelectedTodo] = useState(0);
+
+  const filteredTodos = useMemo(() => {
+    return todos.filter(todo => {
+      switch (filterType) {
+        case FilterType.All:
+          return todo;
+
+        case FilterType.Active:
+          return !todo.completed;
+
+        case FilterType.Completed:
+          return todo.completed;
+
+        default:
+          return true;
+      }
+    });
+  }, [todos, filterType]);
 
   useEffect(() => {
-    // focus the element with `ref={newTodoField}`
-    if (newTodoField.current) {
-      newTodoField.current.focus();
+    inputField.current?.focus();
+
+    const fetchData = async () => {
+      const todosFromServer = await getTodos(user?.id || 0);
+
+      setTodos(todosFromServer);
+    };
+
+    try {
+      fetchData();
+    } catch {
+      setError(Error.LOADING);
     }
   }, []);
 
+  useEffect(() => {
+    inputField.current?.focus();
+  }, [todos]);
+
+  const handleSubmit = useCallback(async (event: FormEvent) => {
+    event.preventDefault();
+
+    if (!title.trim()) {
+      setError(Error.TITLE);
+
+      return;
+    }
+
+    setError(null);
+    setIsAdding(true);
+    setSelectedTodos([user?.id || 0]);
+
+    try {
+      const newTodo = await createTodo(user?.id || 0, title);
+
+      setTodos(prev => [...prev, newTodo]);
+    } catch {
+      setError(Error.ADDING);
+    }
+
+    setTitle('');
+    setIsAdding(false);
+    setSelectedTodos([]);
+  }, [title]);
+
+  const deleteTodo = async (todoId: number) => {
+    setSelectedTodos([todoId]);
+
+    try {
+      await removeTodo(todoId);
+
+      setTodos(todos.filter(todo => todo.id !== todoId));
```

Comment:
If the new state is computed using the previous state, you can pass a function to setTodos